common4j/src/main/com/microsoft/identity/common/java/net/UrlConnectionHttpClient.java

```diff
@@ -0,0 +1,361 @@
+                client.readTimeoutMs,
+                client.connectTimeoutMsSupplier,
+                client.readTimeoutMsSupplier,
+                client.streamBufferSize);
+    }
+
+    /**
+     * Obtain a static default instance of the HTTP Client class.
+     *
+     * @return a default-configured HttpClient.
+     */
+    public static synchronized UrlConnectionHttpClient getDefaultInstance() {
+        UrlConnectionHttpClient reference = defaultReference.get();
+        if (reference == null) {
+            defaultReference.compareAndSet(null, UrlConnectionHttpClient.builder()
+                    .streamBufferSize(STREAM_BUFFER_SIZE)
+                    .retryPolicy(StatusCodeAndExceptionRetry.builder()
+                            .number(1)
+                            .extensionFactor(2)
+                            .isAcceptable(new Function<HttpResponse, Boolean>() {
+                                public Boolean apply(HttpResponse response) {
+                                    return response != null && response.getStatusCode() < 400;
+                                }
+                            })
+                            .initialDelay(RETRY_TIME_WAITING_PERIOD_MSEC)
+                            .isRetryable(new Function<HttpResponse, Boolean>() {
+                                public Boolean apply(HttpResponse response) {
+                                    return response != null && HttpRequest.isRetryableError(response.getStatusCode());
+                                }
+                            })
+                            .isRetryableException(new Function<Exception, Boolean>() {
+                                public Boolean apply(Exception e) {
+                                    return e instanceof SocketTimeoutException;
+                                }
+                            })
+                            .build())
+                    .build());
+            reference = defaultReference.get();
+        }
+        return reference;
+    }
+
+    /**
+     * Record the beginning of an http request.
+     */
+    private static void recordHttpTelemetryEventStart(@NonNull final String requestMethod,
+                                                      @NonNull final URL requestUrl,
+                                                      final String requestId) {
+        Telemetry.emit(
+                new HttpStartEvent()
+                        .putMethod(requestMethod)
+                        .putPath(requestUrl)
+                        .putRequestIdHeader(requestId)
+        );
+    }
+
+    /**
+     * Record the end of an http event.
+     *
+     * @param response
+     */
+    private static void recordHttpTelemetryEventEnd(final HttpResponse response) {
+        final HttpEndEvent httpEndEvent = new HttpEndEvent();
+
+        if (null != response) {
+            httpEndEvent.putStatusCode(response.getStatusCode());
+        }
+
+        Telemetry.emit(httpEndEvent);
```

Comment:
>emit [](http://example.com/codeflow?start=18&length=4)

it will be nice to include requestId here if possible..
So that we can mark the begin and End more precisely by requestId

Reply:
I believe you mean Correlation Id? Telemetry is already aware of Correlation Id - it gets it from DiagnosticContext